README.md
# CNN Hapi

Basic [Hapi](http://hapijs.com/) server with some baked in features that can be
pulled in as a dependency of another application to extend as needed.

[![build](https://img.shields.io/travis/cnnlabs/cnn-hapi/master.svg?style=flat-square)](https://travis-ci.org/cnnlabs/cnn-hapi)
![node](https://img.shields.io/node/v/cnn-hapi.svg?style=flat-square)
[![npm](https://img.shields.io/npm/v/cnn-hapi.svg?style=flat-square)](https://www.npmjs.com/package/cnn-hapi)
[![npm-downloads](https://img.shields.io/npm/dm/cnn-hapi.svg?style=flat-square)](https://www.npmjs.com/package/cnn-hapi)
[![dependency-status](https://gemnasium.com/cnnlabs/cnn-hapi.svg)](https://gemnasium.com/cnnlabs/cnn-hapi)

Features include:

- Swagger on /documentation
- Basic healthcheck on /__health
- Basic logging
- Basic metrics
- Basic process monitoring
- Default Cache/Surrogate Control headers
- Default custom headers


## Requirements

[Node 6.0.0+](https://npmjs.org)


## Installation

```shell
$ npm install
```


## Usage

Look at the [/example/app.js](./example/app.js) to see an example of how this
can be pulled in as a dependency.  You can see it running by doing the
following.

```shell
$ PORT=5000 node example/app.js
info Server running at http://0.0.0.0:5000
info Server name: testHarness
info Server version: 0.1.0
info Server maxListeners: 1000
info Server environment: development
info Server in debug mode: true
160915/014027.438, [ops] memory: 65Mb, uptime (seconds): 5.705, load: [1.30322265625,1.486328125,1.5888671875]
160915/014032.438, [ops] memory: 58Mb, uptime (seconds): 10.706, load: [1.35888671875,1.49462890625,1.5908203125]
```

You can also navigate to localhost:5000 and see a served page.

Swagger documentation - `localhost:5000/documentation`

Healthcheck monitoring - `localhost:5000/__health`

## Testing
`$ cd <cnn-hapi-root>`
`$ npm run example-server`
This will run the example server in the `./example` directory.

## ENV VARS
+ `LOADER_IO_VALIDATION`

+ `PORT`

+ `LOCAL_TLS_PORT`

+ `CACHE_CONTROL`

+ `ENVIRONMENT`

+ `HOST`

+ `DEFAULT_MAX_LISTENERS`

+ `SURROGATE_CACHE_CONTROL`

+ `SHOW_CNN_HAPI_CONFIG`  => Setting this to 'true' will show server instance configurations on `server.start()`. Requires `DEBUG=cnn-hapi*` to be a part of `DEBUG` capture group<br>

+ `METRICS_FLUSHEVERY`<br>


## serverInstance(options)
_The following options set defaults at the server level and can override CNN-Hapi Defaults_
_Populate notes are in order of priority. Example: populate: `process.env.SOMEVALUE` || `options.someValue`. In that example if `process.env.SOMEVALUE` is not set it will default to `options.someValue`, etc, etc_
_Manual override possibilites are expressed in `options.someValue`_

_`options` is an object that can take the following keys_

+ `basePath`: project basePath,

+ `cacheControlHeader`: process.env.CACHE_CONTROL || 'max-age=60',

+ `customHeaders`: options.customHeaders || [],

+ `description`: options.description || `package.json` `description` key,

+ `environment`: process.env.ENVIRONMENT || process.env.NODE_ENV || options.environment || '',

+ `healthChecks`: options.healthChecks || [],

+ `host`: process.env.HOST || options.host || '0.0.0.0',

+ `loaderIoValidationKey`: options.loaderIoValidationKey || undefined,

+ `localTLS`: options.localTLS || null,

+ `maxListeners`: process.env.DEFAULT_MAX_LISTENERS || options.maxListeners || 10,

+ `name`: options.name || `package.json` `name` key,

+ `port`: process.env.PORT || options.port || 3000,

+ `surrogateCacheControl`:
    process.env.SURROGATE_CACHE_CONTROL || options.surrogateCacheControl || 'max-age=360, stale-while-revalidate=60, stale-if-error=86400',

+ `version`:options.version || `package.json` `version` key,

+ `withGoodConsole`: options.withGoodConsole || false,

+ `withSwagger`: options.withSwagger || false

## Override caching on individual routes
_Using the reply.header() function can set headers on a singular route_

```
{
    method: 'GET',
    path: '/override-headers',
    handler: (request, reply) => {
      reply('Peep the response headers in swagger docs')
      .header('Cache-Control', '2')
      .header('Surrogate-Control', 'baz');
    },
    config: {
      description: 'Example route for demonstrating how to  override headers by route',
      tags: ['api']
    }
  },
  ```
_For explicit usage check this implementation in `./example/routes`_


# Test

`$ npm run example-server`

It runs example server located in `./example`. Pass in ENV vars through the above command or hardcode into the package.json located in the `./example` directory.

Go to {HOST}:{PORT}/documentation to view the new swagger docs and to test current `CNN-Hapi` logic.



CHANGELOG.md
# CNN Hapi Changelog

## 2017-08-23, Version 2.5.0, @jennyvallon

### Notable changes

- Add swagger docs
- bug fix: ENV var, server(options), default heirarchy
- bug fix: using reply.header(obj) on `surrogate-control` header
- Add documentation on ENV vars, options and overrides by route implementation
- Add prettier
- ENV var to expose server configs for end user's server instance

### Known issues

See https://github.com/cnnlabs/cnn-hapi/labels/defect for complete and
current list of known issues.

### Commits
* [[`f72a3cbdf7`](https://github.com/cnnlabs/cnn-hapi/commit/f72a3cbdf7)] - Merge pull request #62 from cnnlabs/updateDocumentation (Jenny Vallon)
* [[`c647c6e927`](https://github.com/cnnlabs/cnn-hapi/commit/c647c6e927)] - Merge branch 'master' into develop (Brian Duckett)
* [[`7560b9268c`](https://github.com/cnnlabs/cnn-hapi/commit/7560b9268c)] - Merge branch 'release/2.4.3' (Brian Duckett)

## 2017-07-07, Version 2.3.0, @adslaton

### Notable changes

- Add custom header check

### Known issues

See https://github.com/cnnlabs/cnn-hapi/labels/defect for complete and
current list of known issues.

### Commits
* [[`16ac31235a`](https://github.com/cnnlabs/cnn-hapi/commit/16ac31235a)] - Merge pull request #53 from cnnlabs/fix/cache-headers (A.D. Slaton)
* [[`8c57697c55`](https://github.com/cnnlabs/cnn-hapi/commit/8c57697c55)] - pr feedback: update readme (Brian Duckett)
* [[`91664bd3c1`](https://github.com/cnnlabs/cnn-hapi/commit/91664bd3c1)] - fixes #39 (Brian Duckett)
* [[`8c2eca6e3a`](https://github.com/cnnlabs/cnn-hapi/commit/8c2eca6e3a)] - **fix**: custom header check (Brian Duckett)

## 2017-06-19, Version 2.2.0, @adslaton

### Notable changes

- Update to Hapi 16.4.3

### Known issues

See https://github.com/cnnlabs/cnn-hapi/labels/defect for complete and
current list of known issues.

### Commits

* [[`71920bbd25`](https://github.com/cnnlabs/cnn-hapi/commit/71920bbd25)] - Merge pull request #52 from cnnlabs/feature/hapi-16.4.3 (A.D. Slaton)
* [[`4f4863c828`](https://github.com/cnnlabs/cnn-hapi/commit/4f4863c828)] - fix another travis failure (A.D. Slaton)
* [[`1fdfbc55e4`](https://github.com/cnnlabs/cnn-hapi/commit/1fdfbc55e4)] - fix travis tests (A.D. Slaton)
* [[`36f54ce44b`](https://github.com/cnnlabs/cnn-hapi/commit/36f54ce44b)] - fix travis tests (A.D. Slaton)
* [[`2551007cb0`](https://github.com/cnnlabs/cnn-hapi/commit/2551007cb0)] - fix linting issues (A.D. Slaton)
* [[`91d2b9dbbd`](https://github.com/cnnlabs/cnn-hapi/commit/91d2b9dbbd)] - update hapi to 16.4.3 and update other deps (A.D. Slaton)
* [[`48836a1596`](https://github.com/cnnlabs/cnn-hapi/commit/48836a1596)] - changes logged (Brian Duckett)
* [[`ebd2535523`](https://github.com/cnnlabs/cnn-hapi/commit/ebd2535523)] - Merge branch 'master' into develop (Brian Duckett)
* [[`fb22a71ae6`](https://github.com/cnnlabs/cnn-hapi/commit/fb22a71ae6)] - Merge branch 'release/2.1.1' (Brian Duckett)

## 2017-04-12, Version 2.1.1, @codeclout

### Notable changes

- fixes null metrics provider
- fixes port property

### Known issues

See https://github.com/cnnlabs/cnn-hapi/labels/defect for complete and
current list of known issues.

### Commits

* [[`ebd2535`](https://github.com/cnnlabs/cnn-hapi/commit/ebd2535)] - Merge branch 'master' into develop (Brian Duckett)
* [[`fb22a71`](https://github.com/cnnlabs/cnn-hapi/commit/fb22a71)] - Merge branch 'release/2.1.1' (Brian Duckett)

## 2017-04-07, Version 2.1.0, @adslaton

### Notable changes

- remove layoutsDir from being required

### Known issues

See https://github.com/cnnlabs/cnn-hapi/labels/defect for complete and
current list of known issues.

### Commits

* [[`6e82a09a9d`](https://github.com/cnnlabs/cnn-hapi/commit/6e82a09a9d)] - Merge pull request #49 from cnnlabs/feature/remove-required-layoutsDir (A.D. Slaton)
* [[`74530300a5`](https://github.com/cnnlabs/cnn-hapi/commit/74530300a5)] - Merge branch 'develop' into feature/remove-required-layoutsDir (A.D. Slaton)
* [[`1ec1bc47f2`](https://github.com/cnnlabs/cnn-hapi/commit/1ec1bc47f2)] - remove layoutsDir from being required (A.D. Slaton)
* [[`79d3b29f25`](https://github.com/cnnlabs/cnn-hapi/commit/79d3b29f25)] - changelog (Brian Duckett)
* [[`19091e5d51`](https://github.com/cnnlabs/cnn-hapi/commit/19091e5d51)] - Merge branch 'master' into develop (Brian Duckett)
* [[`9d33506051`](https://github.com/cnnlabs/cnn-hapi/commit/9d33506051)] - Merge branch 'release/2.0.3' (Brian Duckett)

## 2017-04-07, Version 2.0.3, @codeclout

### Notable changes

- hapiJS dependency updated to v16.1.1

### Known issues

See https://github.com/cnnlabs/cnn-hapi/labels/defect for complete and
current list of known issues.

### Commits

* [[`19091e5d51`](https://github.com/cnnlabs/cnn-hapi/commit/19091e5d51)] - Merge branch 'master' into develop (Brian Duckett)
* [[`9d33506051`](https://github.com/cnnlabs/cnn-hapi/commit/9d33506051)] - Merge branch 'release/2.0.3' (Brian Duckett)

## 2017-03-16, Version 2.0.2, @codeclout

### Notable changes

- N/A

### Known issues

See https://github.com/cnnlabs/cnn-hapi/labels/defect for complete and
current list of known issues.

### Commits

* [[`329ca08`](https://github.com/cnnlabs/cnn-hapi/commit/329ca08)] - Merge branch 'release/2.0.2' (Brian Duckett)

## 2017-03-16, Version 2.0.1, @codeclout

### Notable changes

- N/A

### Known issues

See https://github.com/cnnlabs/cnn-hapi/labels/defect for complete and
current list of known issues.

### Commits

* [[`cfcf77e`](https://github.com/cnnlabs/cnn-hapi/commit/cfcf77e)] - Merge branch 'release/2.0.1' (Brian Duckett)

## 2017-03-16, Version 2.0.0, @codeclout

### Notable changes

- hapiJS dependency updated to v16.1.0
- release tool fix
- moves socket.io & handlebars to example app

### Known issues

See https://github.com/cnnlabs/cnn-hapi/labels/defect for complete and
current list of known issues.

### Commits

* [[`cbe3605d87`](https://github.com/cnnlabs/cnn-hapi/commit/cbe3605d87)] - release tool tweak (Brian Duckett)

lib/config.js
const Fs = require('fs');
const Joi = require('joi');
const path = require('path');

const debug = require('debug')(`cnn-hapi:${path.basename(__filename)}`);

const {resolve} = require('path');

class Config {
  constructor({description, name, version}, options, basePath) {
    const internals = {
      defaults: {
        basePath: basePath,
        cacheControlHeader: process.env.CACHE_CONTROL || 'max-age=60',
        customHeaders: options.customHeaders || [],
        description: options.description || description,
        environment: process.env.ENVIRONMENT || process.env.NODE_ENV || options.environment || '',
        healthChecks: options.healthChecks || [],
        host: process.env.HOST || options.host || '0.0.0.0',
        loaderIoValidationKey: options.loaderIoValidationKey || undefined,
        localTLS: options.localTLS || null,
        maxListeners: process.env.DEFAULT_MAX_LISTENERS || options.maxListeners || 10,
        name: options.name || name,
        port: process.env.PORT || options.port || 3000,
        surrogateCacheControl:
          process.env.SURROGATE_CACHE_CONTROL ||
          options.surrogateCacheControl ||
          'max-age=360, stale-while-revalidate=60, stale-if-error=86400',
        version: version,
        withGoodConsole: options.withGoodConsole || false,
        withSwagger: options.withSwagger || false
      }
    };

    if (process.env.SHOW_CNN_HAPI_CONFIG != undefined && process.env.SHOW_CNN_HAPI_CONFIG === 'true') {
      debug('Server configurations %o', internals.defaults);
    }

    internals.schema = {
      basePath: Joi.string()
        .description('directory of the current process')
        .example('/Users/mjr'),
      cacheControlHeader: Joi.string()
        .description('maximum cache lifetime in seconds')
        .example('max-age=60'),
      customHeaders: Joi.array(),
      description: Joi.string()
        .description('a description of the application')
        .example('CNN Hapi'),
      environment: Joi.string()
        .description('application environment')
        .example('dev'),
      healthChecks: Joi.array(),
      host: Joi.string()
        .description('application host')
        .example('0.0.0.0'),
      loaderIoValidationKey: Joi.string().description('loader.io validation'),
      layoutsDir: Joi.string()
        .description('the application view')
        .example('`${__dirname}/views/`'),
      localTLS: Joi.object()
        .keys({
          cert: Joi.string(),
          key: Joi.string(),
          port: Joi.number().integer()
        })
        .allow(null),
      maxListeners: Joi.number()
        .integer()
        .description('max listeners per event')
        .example(1000),
      metrics: Joi.object().keys({
        provider: Joi.object(),
        options: Joi.object()
      }),
      name: Joi.string()
        .description('application name')
        .example('cnn-hapi'),
      port: Joi.number()
        .integer()
        .description('http port')
        .example(3000),
      surrogateCacheControl: Joi.string()
        .description('external cache layer')
        .example('max-age=360, stale-while-revalidate=60, stale-if-error=86400'),
      version: Joi.string()
        .description('version of the application')
        .example('0.0.1'),
      withGoodConsole: Joi.boolean(),
      withSwagger: Joi.boolean()
    };

    this.settings = Object.assign(internals.defaults, {description, name, version});
    Joi.assert(this.settings, internals.schema, 'Invalid service configuration');

    this.settings.isProduction = () => {
      const {environment} = this.settings;

      return environment.toLowerCase() === 'production' || environment.toLowerCase() === 'prod';
    };
  }

  get httpConnection() {
    const {host, labels, port} = this.settings;

    return {
      host,
      labels: labels || 'http',
      port
    };
  }

  get httpsConnection() {
    const {cert, key, port = 3000} = this.settings.localTLS;

    const {basePath, host, labels} = this.settings;

    return {
      host,
      labels: Array.isArray(labels) ? [...labels, 'local-tls'] : 'local-tls',
      port,
      routes: {security: {hsts: true}},
      tls: {
        key: Fs.readFileSync(resolve(basePath, `./${key}`)),
        cert: Fs.readFileSync(resolve(basePath, `./${cert}`))
      }
    };
  }
}

module.exports = Config;
